[lldb] Provide lr value in faulting frame on arm64

jasonmolenda opened the pull request:
When a frameless function faults or is interrupted asynchronously, the UnwindPlan MAY have no register location rule for the return address register (lr on arm64); the value is simply live in the lr register when it was interrupted, and the frame below this on the stack -- e.g. sigtramp on a Unix system -- has the full register context, including that register.

RegisterContextUnwind::SavedLocationForRegister, when asked to find the caller's pc value, will first see if there is a pc register location.  If there isn't, on a Return Address Register architecture like arm/mips/riscv, we rewrite the register request from "pc" to "RA register", and search for a location.

On frame 0 (the live frame) and an interrupted frame, the UnwindPlan may have no register location rule for the RA Reg, that is valid. A frameless function that never calls another may simply keep the return address in the live register the whole way.  Our instruction emulation unwind plans explicitly add a rule (see Pavel's May 2024 change https://github.com/llvm/llvm-project/pull/91321 ), but an UnwindPlan sourced from debug_frame may not.

I've got a case where this exactly happens - clang debug_frame for arm64 where there is no register location for the lr in a frameless function.  There is a fault in the middle of this frameless function and we only get the lr value from the fault handler below this frame if lr has a register location of `IsSame`, in line with Pavel's 2024 change.

Similar to how we see a request of the RA Reg from frame 0 after failing to find an unwind location for the pc register, the same style of special casing is needed when this is a function that was interrupted.

Without this change, we can find the pc of the frame that was executing when it was interrupted, but we need $lr to find its caller, and we don't descend down to the trap handler to get that value, truncating the stack.

rdar://145614545

jasonmolenda committed:
[lldb] Provide lr value in faulting frame on arm64

When a frameless function faults or is interrupted asynchronously,
the UnwindPlan MAY have no register location rule for the return
address register (lr on arm64); the value is simply live in the
lr register when it was interrupted, and the frame below this on
the stack -- e.g. sigtramp on a Unix system -- has the full
register context, including that register.

RegisterContextUnwind::SavedLocationForRegister, when asked to find
the caller's pc value, will first see if there is a pc register
location.  If there isn't, on a Return Address Register architecture
like arm/mips/riscv, we rewrite the register request from "pc" to
"RA register", and search for a location.

On frame 0 (the live frame) and an interrupted frame, the UnwindPlan
may have no register location rule for the RA Reg, that is valid.
A frameless function that never calls another may simply keep the
return address in the live register the whole way.  Our instruction
emulation unwind plans explicitly add a rule (see Pavel's May 2024
change https://github.com/llvm/llvm-project/pull/91321 ), but an
UnwindPlan sourced from debug_frame may not.

I've got a case where this exactly happens - clang debug_frame for
arm64 where there is no register location for the lr in a frameless
function.  There is a fault in the middle of this frameless function
and we only get the lr value from the fault handler below this frame
if lr has a register location of `IsSame`, in line with Pavel's
2024 change.

Similar to how we see a request of the RA Reg from frame 0 after
failing to find an unwind location for the pc register, the same
style of special casing is needed when this is a function that
was interrupted.

Without this change, we can find the pc of the frame that was
executing when it was interrupted, but we need $lr to find its
caller, and we don't descend down to the trap handler to get that
value, truncating the stack.

rdar://145614545

jasonmolenda commented:
While working on this bug, I started musing about how we could switch the logic for how to handle registers more into UnwindPlans, wrote a little discourse with the idea.  https://discourse.llvm.org/t/unhappiness-with-the-lldb-unwinder-register-passing-up-the-stack-interrup-trap-sigtramp-frames/86058

It's not something I'll be doing in the near term, but I know I'd like to get back to this.

jasonmolenda commented:
Also while initially working on this issue I found I could fix it in two places in `SavedLocationForRegister` without considering the entirety of the method.  This made me unhappy so I spent a bit of time going over all of `SavedLocationForRegister` until I felt confident I understood it, and then wrote the scariest NFC patch where I removed chunks of it (that are doing nothing) and restructured it so it was a lot easier to read, I thought.  But I want to maybe bring that up separately because it's a lot of change and this method has a lot of edge cases - many tricky to test, and not tested well.

jasonmolenda commented:
to be clear, when I mention a big chunk of `SavedLocationForRegister` which I don't think is actually doing anything -- I refuse to look at the history, but it looks suspiciously like something I wrote a decade ago :)

jasonmolenda commented:
I am also not thrilled with the GetAsKind method for register numbering returning LLDB_INVALID_REGNUM, the code reads so poorly in this method because of it.  But returning an optional would make other code more verbose unless I carried optionals through a lot further.  I haven't been annoyed enough to try doing that yet.

labath commented:
This makes sense to me, but is there any way to write a test case for it?

It sounds like something similar to [this test](https://github.com/llvm/llvm-project/blob/main/lldb/test/Shell/Unwind/signal-in-leaf-function-aarch64.test) could work. In fact, given that this test is disabled on darwin, I'm wondering if this patch does not fix it by any chance?

jasonmolenda committed:
Merge branch 'main' into fetch-lr-for-frameless-function-async-interrupted

jasonmolenda commented:
FTR that test is skipped on darwin because _sigtramp in libc on aarch64 doesn't have any hand-supplied eh_frame right now :/

I have a test case put together, with a few functions written in assembly with hand-added cfi directives.  It's probably darwin specific assembly, obviously aarch64 only, but it exercises exactly the bug I am fixing.  will wrap it up into a proper test in a tiny bit (hopefully tonight)

jasonmolenda committed:
Add the sources for an API test case to be written

jasonmolenda committed:
Add a log msg when a frame is marked as a trap handler
via the UnwindPlan.  And use the trap handler flag for
zeroth frame so we can properly unwind a frameless function
that was interrupted while in the trap handler function.

jasonmolenda committed:
Update assembly a bit.

jasonmolenda commented:
I'm thinking I'll write an API test that instruction steps through the whole thing and backtraces at each point, making sure it can get every function between the current one & main at all points, just for fun extra stress testing.

jasonmolenda committed:
Add unwind rules for x0 (volatile) and x20 (non-voltile)
to trap() and break_to_debugger() for fun, and add
unwind instructions for the epilogue of trap().

When stopped in `break_to_debugger`,
```
  * frame #0: 0x00000001000003e8 a.out`break_to_debugger + 4
    frame #1: 0x00000001000003d8 a.out`trap + 16
    frame #2: 0x00000001000003c0 a.out`to_be_interrupted + 20
    frame #3: 0x0000000100000398 a.out`main + 32
```

Normally we can't provide a volatile register (e.g. x0) up
the stack.  And we can provide a non-volatile register (e.g. x20)
up the stack.  I added an IsSame rule for trap() and break_to_debugger()
for x0, so it CAN be passed up the stack.  And I added an Undefined
rule for x20 to trap() so it CAN'T be provided up the stack.

If a user selects `to_be_interrupted` and does `register read`,
they'll get x0 and they won't get x20.  From `main`, they will not
get x0 or x20 (x0 because `to_be_interrupted` doesn't have an IsSame
rule).

jasonmolenda committed:
small comment improvement

jasonmolenda committed:
more comment fix

jasonmolenda committed:
Add test case Makefile and API test python code.
Also fix one bug in the by-hand unwind state in
the assembly file.

jasonmolenda committed:
Add tests for x0 and x20, which I treat in normally ABI-wrong
ways in these hand-written UnwindPlans.

jasonmolenda committed:
Remove some Mach-O/Darwinisms, get closer to
this compiling on linux?
I think maybe the only difference is that the
symbol names are prefixed with _ on darwin?

jasonmolenda commented:
OK finished writing the test case.  It starts at main() and instruction steps through the test program, checking that it can backtrace all the way to main() (including all the frames in the middle) at each instruction.  

I also added tests for the non-ABI compliant fetching of x0 mid-stack and x20 being unfetchable mid-stack, with the hand-written eh_frame instructions in the assembly file.  It has nothing to do with the frameless function that faults that I'm trying to fix here, but it was interesting that it works as I wanted so I'll add a test on it.

I have the test marked as skipUnlessDarwin but I'm not sure there's much to keep it from working on Linux.  Maybe that I had to prefix the symbol names with _ on dawin?  Might be easiest to update the .c file to call either to_be_interrupted or _to_be_interrupted depending on the OS if that's it.

JDevlieghere commented on lldb/test/API/macosx/unwind-frameless-faulted/Makefile:
```diff
@@ -0,0 +1,12 @@
+C_SOURCES := main.c
+
+interrupt-and-trap-funcs.o: interrupt-and-trap-funcs.s
+	$(CC) $(CFLAGS) -c -o interrupt-and-trap-funcs.o $(SRCDIR)/interrupt-and-trap-funcs.s
```
Should this pass a triple? I assume this works locally because you're on an arm64 host?

JDevlieghere reviewed (approved):
Thanks for taking the time to com up with this test case. Pavel already signed off on the implementation so I think this is good to merge.

jasonmolenda committed:
Adapt test case to build & run on linux or
Darwin.

There's a slight syntax diff for the ADRP + ADD pair and the label
name.  And the function names in Darwin are prepended with an _ and
not on linux.

clang on darwin runs the .s file through the preprocessor, but it
doesn't seem to on linux.  I renamed interrupt-and-trap-funcs.s to
interrupt-and-trap-funcs.c and run it through the preprocessor then
assemble it in Makefile now.

The linux version would probably run on other AArch64 systems like
FreeBSD but I haven't checked those.

jasonmolenda committed:
Move this test from macosx/ to functionalities/unwind
because it can run on darwin or linux aarch64.

jasonmolenda commented:
Minor syntax adaptations to the assembly file to build on an aarch64 unbuntu 24.04, test updated and will run on linux or darwin now.

jasonmolenda committed:
typeo

jasonmolenda committed:
Disabling clang-format of the assembly file called ".c"

Add two more checks for non-ABI compliant register availability
from the custom UnwindPlans.

jasonmolenda commented:
Ah, lldb-server doesn't secretly migrate the pc past a builtin_debugtrap() BRK instruction on linux like debugserver does, the test will inf loop never advancing past the BRK.

jasonmolenda committed:
Advance past the BRK instruction in break_to_debugger() for lldb-server
where it does not advance past this BRK instruction automatically.

Also add a limit on instruction stepping, so if this does get stuck
in a loop it won't run forever.

I haven't tested this on a linux system yet but in by-hand running
on the binary, I saw the lldb-server behavior w.r.t builtin_debugtrap
and this should be necessary.

jasonmolenda committed:
typeo

jasonmolenda committed:
Shift the underscore-versus-not assembly naming convention into the
.s file -- the Test python file also hardcodes the names of the
functions and I don't want to prepend _ on linux to those names.

jasonmolenda commented:
I had two commits to address bots.  The first was my skipIf to run the test only darwin/linux only for arm64/aarch64 acted like an or so it could run on either of those.  Second, the aarch64-ubuntu bot failed when I had it my assembly file named ".c" and did `clang -E -o interrupt-and-trap-funcs.s interrupt-and-trap-funcs.c` to run it through the preprocessor - clang on that system when run by the API tests injected a bunch of C headers and it failed to compile as assembly then.  

For now, I changed the test to only run on Darwin.

jasonmolenda commented:
I reverted it entirely when the lldb incremental arm64 bot failed the test.  On that older Xcode/OS (Xcode 15.2, macOS 14.1) there was something wrong with the CFI directives and the backtrace out of `trap()` failed.  On a modern macOS system this works fine.  I'll set up an older system and repo.

labath commented:
I looked at why the test fails on linux. It turns out it's due to a bunch of reasons, but most of them are unrelated to the problem at hand. #139545 and #139550 fix the surrounding issues, and https://github.com/llvm/llvm-project/commit/c290e555d6e7d196ebbe5fd6d64e0be25275a5b4 is enough to fix the test itself.

I needed to change the compile command because, for some reason the compiler was not finding the temporary file. I didn't look at what causes the difference because I think this is more robust and idiomatic.

I also needed to remove the `brk #0xf000` instruction because lldb is not able to step over it -- it just ends up spinning forever. FWICS, it's not actually necessary now that you're stepping through the test. Nonetheless, this is a bug -- one that @DavidSpickett might be interested in :)

Interestingly, the mac debug server does not step onto the breakpoint instruction at all. It skips over it when stepping over the previous instruction (i.e. that single step operation actually steps two instructions). This is most likely also some kind of a bug.

> FTR that test is skipped on darwin because _sigtramp in libc on aarch64 doesn't have any hand-supplied eh_frame right now :/

Oh... FWIW, we're in the same situation on aarch64 linux, but here we hard code the plan into lldb in `PlatformLinux::GetTrapHandlerUnwindPlan` (which we can, because it's part of the OS ABI). Maybe you could do the same?

DavidSpickett commented:
> I also needed to remove the brk #0xf000 instruction because lldb is not able to step over it -- it just ends up spinning forever. FWICS, it's not actually necessary now that you're stepping through the test. Nonetheless, this is a bug -- one that @DavidSpickett might be interested in :)

Yes I've been asked a few times about it. GDB also has problems with it, but probably does something slightly more useful. I'll speak to folks on that side and see how we could handle this.

jasonmolenda commented:
> > I also needed to remove the brk #0xf000 instruction because lldb is not able to step over it -- it just ends up spinning forever. FWICS, it's not actually necessary now that you're stepping through the test. Nonetheless, this is a bug -- one that @DavidSpickett might be interested in :)
> 
> Yes I've been asked a few times about it. GDB also has problems with it, but probably does something slightly more useful. I'll speak to folks on that side and see how we could handle this.

@DavidSpickett sorry I missed this discussion a few weeks ago.  Yeah, in debugserver aarch64 backend we check for the current instruction being `brk #0xf000` specifically and move $pc to the next instruction.  Customers were using `__builtin_debugtrap()` in their code to force a break in the debugger, and wanted to continue stepping after hitting that, like they could on Intel systems (where the pc was past the interrupt instruction).  I wasn't sure if it was best to special case this in debugserver, or up in lldb in some architecture dependent stepping plan when we're stopped at this instruction, I may have chosen poorly.
